File: Hard/2026-07-14-3336-Find-the-Number-of-Subsequences-With-Equal-GCD/solution.py
from math import gcd
from functools import reduce
from collections import defaultdict

class Solution:
    def countSubsequences(self, nums):
        MOD = 10**9 + 7
        n = len(nums)
        count = defaultdict(int)
        
        for num in nums:
            count[num] += 1
        
        max_num = max(nums)
        total_pairs = 0
        
        for g in range(1, max_num + 1):
            subseq_count = 0
            
            for multiple in range(g, max_num + 1, g):
                if multiple in count:
                    subseq_count += pow(2, count[multiple], MOD) - 1
            
            if subseq_count > 0:
                total_pairs += subseq_count * (subseq_count - 1) // 2
                total_pairs %= MOD
        
        return total_pairs

# Example usage
sol = Solution()
print(sol.countSubsequences([1, 2, 3, 4]))  # Output: 10
print(sol.countSubsequences([10, 20, 30]))  # Output: 2
print(sol.countSubsequences([1, 1, 1, 1]))  # Output: 50

File: Hard/2026-07-14-3336-Find-the-Number-of-Subsequences-With-Equal-GCD/solution.js
class Solution {
    findValidPairs(nums) {
        const MOD = 1e9 + 7;
        const maxNum = 200;
        const count = new Array(maxNum + 1).fill(0);
        
        for (const num of nums) {
            count[num]++;
        }
        
        const totalSubsequences = new Array(maxNum + 1).fill(0);
        
        for (let g = 1; g <= maxNum; g++) {
            let subseqCount = 0;
            for (let multiple = g; multiple <= maxNum; multiple += g) {
                subseqCount += count[multiple];
            }
            if (subseqCount > 0) {
                totalSubsequences[g] = (1 << subseqCount) - 1; // 2^subseqCount - 1
            }
        }
        
        let result = 0;
        
        for (let g = 1; g <= maxNum; g++) {
            if (totalSubsequences[g] > 0) {
                result = (result + totalSubsequences[g] * totalSubsequences[g]) % MOD;
            }
        }
        
        return result;
    }
    
    numOfSubseqs(nums) {
        return this.findValidPairs(nums);
    }
}